framework/src/utils/MooseCoordTransform.C

```diff
@@ -0,0 +1,338 @@
+                        (_coord_type == Moose::COORD_RSPHERICAL && _z_axis == INVALID),
+                    "'_z_axis' is not an expected value");
+        _r_axis = X;
+        _z_axis = _coord_type == Moose::COORD_RZ ? Z : INVALID;
+      }
+      else if (_r_axis == Y)
+      {
+        mooseAssert((_coord_type == Moose::COORD_RZ && _z_axis == X) ||
+                        (_coord_type == Moose::COORD_RSPHERICAL && _z_axis == INVALID),
+                    "'_z_axis' is not an expected value");
+        _r_axis = Z;
+        _z_axis = _coord_type == Moose::COORD_RZ ? X : INVALID;
+      }
+      axes_rotated = true;
+    }
+  }
+
+  _rotate = std::make_unique<RealTensorValue>(
+      RealTensorValue::extrinsic_rotation_matrix(alpha, beta, gamma));
+
+  if (must_rotate_axes && !axes_rotated)
+    mooseError("Unsupported manual angle prescription in 'MooseCoordTransform::setRotation'");
+}
+
+void
+MooseCoordTransform::setLengthUnitsPerMeter(const Real length_units_per_meter)
+{
+  const auto scale = 1 / length_units_per_meter;
+
+  _scale =
+      std::make_unique<RealTensorValue>(RealTensorValue(scale, 0, 0, 0, scale, 0, 0, 0, scale));
+}
+
+void
+MooseCoordTransform::setCoordinateSystem(const Moose::CoordinateSystemType coord_type,
+                                         const Direction rz_symmetry_axis)
+{
+  _coord_type = coord_type;
+
+  if (_coord_type == Moose::COORD_RZ)
+  {
+    if (rz_symmetry_axis == INVALID)
+      mooseError("For RZ coordinate systems, the 'rz_symmetry_axis' parameter must be provided to "
+                 "'MooseCoordTransform::setCoordinateSystem'");
+
+    _z_axis = rz_symmetry_axis;
+    _r_axis = _z_axis == X ? Y : X;
+  }
+  else if (_coord_type == Moose::COORD_RSPHERICAL)
+    _r_axis = X;
+}
+
+InputParameters
+MooseCoordTransform::validParams()
+{
+  auto params = emptyInputParameters();
+  /// One entry of coord system per block, the size of _blocks and _coord_sys has to match, except:
+  /// 1. _blocks.size() == 0, then there needs to be just one entry in _coord_sys, which will
+  ///    be set for the whole domain
+  /// 2. _blocks.size() > 0 and no coordinate system was specified, then the whole domain will be XYZ.
+  /// 3. _blocks.size() > 0 and one coordinate system was specified, then the whole domain will be that system.
+  params.addParam<std::vector<SubdomainName>>("block", "Block IDs for the coordinate systems");
+  MultiMooseEnum coord_types("XYZ RZ RSPHERICAL", "XYZ");
+  MooseEnum rz_coord_axis("X=0 Y=1", "Y");
+  params.addParam<MultiMooseEnum>(
+      "coord_type", coord_types, "Type of the coordinate system per block param");
+  params.addParam<MooseEnum>(
+      "rz_coord_axis", rz_coord_axis, "The rotation axis (X | Y) for axisymetric coordinates");
+  params.addParam<Real>("length_units_per_meter",
+                        "How many mesh length units are in a meter, e.g. if your mesh units are "
+                        "centimeters, then this parameter value should be 100.");
```

Comment:
```suggestion
  params.addParam<std::string>("length_units",
                        "Length units of the mesh, e.g cm, nm, ft, 5in.");
```

Reply:
also link to Unit system the coord transfrom documentation in syntax/transfers page pls